[ROCm] Bazel build and continuous integration infrastructure

whchung opened the pull request:
This pull request is to start introduce support for ROCm platform to TensorFlow. In this initial pull request, 2 components are addressed:

- bazel build system
- continuous integration logic

Authors:

- Jack Chung: jack.chung@amd.com
- Jeffrey Poznanovic: Jeffrey.Poznanovic@amd.com
- Peng Sun: Peng.Sun@amd.com

whchung committed:
[ROCm] bazel build system and continuous integration logic

The commit contains following components to support TensorFlow on ROCm platform

- bazel build system
- continuous integration logic

Authors:

- Jack Chung: jack.chung@amd.com
- Jeffrey Poznanovic: Jeffrey.Poznanovic@amd.com
- Peng Sun: Peng.Sun@amd.com

whchung commented:
ping?

yifeif reviewed (dismissed):
configure and build script change lgtm

caisq commented on tensorflow/tools/ci_build/Dockerfile.rocm:
```diff
@@ -0,0 +1,98 @@
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+# Workaround: use HIP PR#457 and then build from source
```
Probably better to provide the full link here for posterity.

whchung replied:
`HIP` PR#457 ( https://github.com/ROCm-Developer-Tools/HIP/pull/457 ) had been merged into `HIP` mainline after this PR was created. I'll amend this PR to address this fact.

/cc @paralleo for awareness

whchung replied:
@caisq pushed new commit to address this

caisq commented on configure.py:
```diff
@@ -1492,6 +1492,13 @@ def main():
       set_trisycl_include_dir(environ_cp)
 
+  set_action_env_var(environ_cp, 'TF_NEED_ROCM', 'ROCm', False)
+  if environ_cp.get('TF_NEED_ROCM') == '1':
```
nit: These two `if` statements can be consolidated as one, because neither of them has an `else` branch.

whchung replied:
I'd like to keep them as is for now. There are additional upcoming ROCm-specific checks / env vars in future PRs. For example, like in CUDA path where it's possible to build TensorFlow with either `nvcc` or `CUDA clang`, we are also working on similar route where it's possible to switch between incumbent `HIP/HCC` toolchain or upcoming `HIP clang` toolchain.

whchung replied:
@gunan for the time being `HIP/HCC` would be the incumbent toolchain on ROCm and we'll switch to `HIP clang` toolchain once its performance beats the incumbent solution.

I'm working on revising this PR to address comments from all reviewers now and will ping you once it's ready.

caisq commented on configure.py:
```diff
@@ -1492,6 +1492,13 @@ def main():
     else:
       set_trisycl_include_dir(environ_cp)
 
+  set_action_env_var(environ_cp, 'TF_NEED_ROCM', 'ROCm', False)
```
`TF_NEED_CUDA`, `TF_NEED_SYCL` and `TF_NEED_ROCM` are all mutually exclusive, right? If so, we need a sanity check here that at most one of them is true.

whchung replied:
From TensorFlow perspective they are not mutually exclusive. It's possible to enable all three and still have TensorFlow built.

whchung replied:
ok will do.

whchung replied:
revised the PR to raise error in case more than 2 GPU platforms are specified (CUDA/SYCL/ROCm).

gunan commented on tensorflow/tensorflow.bzl:
```diff
@@ -910,28 +921,50 @@ def _cuda_copts():
                           hdrs=[],
                           **kwargs):
-  copts = copts + _cuda_copts() + if_cuda(cuda_copts) + tf_copts()
+  copts=copts + tf_copts() + _cuda_copts() + _rocm_copts() + if_cuda_is_configured(cuda_copts) + if_rocm_is_configured(cuda_copts)
```
let's hide the additional complexities here by adding an arg to "_cuda_copts" and "_rocm_copts"

So this line becomes:
copts=copts + tf_copts() + _cuda_copts(opts=cuda_copts) + _rocm_copts(opts=cuda_copts)

whchung replied:
@gunan Thanks. I'll address this.

gunan commented on tensorflow/tools/ci_build/Dockerfile.rocm:
```diff
@@ -0,0 +1,97 @@
+# This Dockerfile provides a starting point for a ROCm installation of 
+# MIOpen and tensorflow.  
+FROM ubuntu:xenial-20170619
```
Why not a more generic image? "ubuntu:xenial"

whchung replied:
@parallelo Please weigh in. I believe the specific tag was specified to ensure the proper Linux kernel version is used?

parallelo replied:
Correct.  We typically use a 4.13-45 linux kernel.

gunan replied:
Do we have an upper limit on the kernel version?
If we depend on one very specific kernel version this will be very very brittle.
Is there a way to lift this restriction?

whchung replied:
we had an issue with 4.15+ kernels in the older version of ROCm stack. Now the issue has been fixed in the upcoming ROCm release so I'll remove this.

gunan commented on tensorflow/tools/ci_build/builds/with_the_same_user:
```diff
@@ -48,6 +48,12 @@ getent passwd "${CI_BUILD_UID}" || adduser ${ADDUSER_OPTS} \
+if [[ "${TF_NEED_ROCM}" -eq 1 ]]; then
+  # ROCm requires the video group in order to use the GPU for compute. If it
+  # exists on the host, add it to the container.
+  getent group video || addgroup video && adduser "${CI_BUILD_USER}" video
```
@caisq I am quite unfamiliar with this script. Are the changes here OK?

whchung replied:
@parallelo This line is to fulfill permission requirements from ROCm stack specified at: https://github.com/RadeonOpenCompute/ROCm#next-set-your-permissions

parallelo replied:
In the case of ROCm, we add the container's user to the `video` group (but only if the host user was a member of the `video` group).  This group membership is currently a requirement for ROCm.

gunan commented on tensorflow/tools/ci_build/linux/rocm/run_cc_core.sh:
```diff
@@ -0,0 +1,40 @@
+bazel test --config=rocm --test_tag_filters=-no_oss,-oss_serial,-no_gpu,-benchmark-test -k \
+    --test_lang_filters=cc --jobs=${N_JOBS} --test_timeout 300,450,1200,3600 \
+    --build_tests_only --test_output=errors --local_test_jobs=1 --config=opt \
+    --run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute -- \
```
I recommend omitting this flag, and setting local_test_jobs=1.
This is highly specialized for VMs with 8 k80 GPUs attached.

whchung replied:
thanks. will do.

gunan commented on tensorflow/tools/ci_build/linux/rocm/run_py3_core.sh:
```diff
@@ -0,0 +1,40 @@
+bazel test --config=rocm --test_tag_filters=-no_oss,-oss_serial,-no_gpu,-benchmark-test -k \
+    --test_lang_filters=py --jobs=${N_JOBS} --test_timeout 300,450,1200,3600 \
+    --build_tests_only --test_output=errors --local_test_jobs=1 --config=opt \
+    --run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute -- \
```
ditto above. remove this option.

whchung replied:
thanks. will do.

gunan commented on tensorflow/tools/ci_build/osx/libtensorflow_rocm.sh:
```diff
@@ -0,0 +1,37 @@
+
+# Configure script
+export TF_NEED_ROCM=1
+export LD_LIBRARY_PATH="/usr/local/cuda/lib:/usr/local/cuda/extras/CUPTI/lib:${LD_LIBRARY_PATH}"
```
You can remove this cuda specific line.

whchung replied:
thanks. will do.

gunan commented on tensorflow/tools/ci_build/xla/linux/rocm/run_py3.sh:
```diff
@@ -0,0 +1,42 @@
+bazel test --config=rocm --test_tag_filters=-no_gpu,-benchmark-test,-no_oss -k \
+    --jobs=${N_JOBS} --test_timeout 300,450,1200,3600 \
+    --build_tests_only --test_output=errors --local_test_jobs=1 \
+    --run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute \
```
remove this option.

whchung replied:
thanks. will do.

gunan commented on third_party/gpus/crosstool/CROSSTOOL_hipcc.tpl:
```diff
@@ -0,0 +1,277 @@
+major_version: "local"
```
@meteorcloudy @mhlopko could you help review this file?

gunan commented on third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_rocm.tpl:
```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python
```
@meteorcloudy @mhlopko could you review?

gunan commented on third_party/gpus/rocm/BUILD.tpl:
```diff
@@ -0,0 +1,128 @@
+    },
+)
+
+config_setting(
```
I do not think we need these duplicated here. Is there a reason to not reuse them from //tensorflow/BUILD?

whchung replied:
@gunan configs here are implemented following `tensorflow/third_party/gpus/cuda/BUILD.tpl` where CUDA-specific configs are specified. These CUDA-specific configs aren't in `//tensorflow/BUILD` either.

Since the purpose of this PR is to introduce ROCm-specific build scripts without major refactoring to existing TensorFlow build infrastructure I'd like to propose keep this file as is. Thoughts?

gunan commented on third_party/gpus/rocm/build_defs.bzl.tpl:
```diff
@@ -0,0 +1,32 @@
+# Macros for building ROCm code.
```
@meteorcloudy @mhlopko could you help review?

meteorcloudy commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+  cc = find_cc(repository_ctx)
+  host_compiler_includes = _host_compiler_includes(repository_ctx, cc)
+  rocm_defines = {
+           "%{rocm_include_path}": _rocm_include_path(repository_ctx,
```
Looks like you have hard-coded rocm include paths in CROSSTOOL, should we remove this field or not hard-code?

whchung replied:
I'll migrate hard-coded paths from CROSSTOOL to `rocm_configure.bzl` so it's easier to maintain. In CROSSTOOL we'll honor `%{rocm_include_path}`.

meteorcloudy commented on third_party/gpus/crosstool/CROSSTOOL_hipcc.tpl:
```diff
@@ -0,0 +1,277 @@
+  # linker_flag: "-Wl,--detect-odr-violations"
+
+  # Include directory for ROCm headers.
+  cxx_builtin_include_directory: "/opt/rocm/hsa/include"
```
Should we replace them with `%{rocm_include_path}` so that they can be configured?

whchung replied:
thanks. let me see how to remove these `cxx_builtin_include_directory` and have them populated from `rocm_configure.bzl`.

perfinion commented on tensorflow/workspace.bzl:
```diff
@@ -35,6 +36,7 @@ def tf_workspace(path_prefix="", tf_repo_name=""):
   nccl_configure(name="local_config_nccl")
   git_configure(name="local_config_git")
   sycl_configure(name="local_config_sycl")
+  rocm_configure(name="local_config_rocm")
```
You probably need to add an exclude to build_pip_package.sh for this too.
here: https://github.com/tensorflow/tensorflow/blob/622a111e88dfbac430556ec8213230345189f466/tensorflow/tools/pip_package/build_pip_package.sh#L30

rmlarsen commented:
@gunan could you please take another look?

whchung commented on configure.py:
```diff
@@ -1492,6 +1492,13 @@ def main():
+  if environ_cp.get('TF_NEED_ROCM') == '1':
+    if 'LD_LIBRARY_PATH' in environ_cp and environ_cp.get(
+        'LD_LIBRARY_PATH') != '1':
+      write_action_env_to_bazelrc('LD_LIBRARY_PATH',
```
`TF_NEED_ROCM` and `TF_NEED_CUDA` would be changed so they are mutually exclusive.

whchung commented:
@rmlarsen / @gunan please hold it for now . I haven't finished addressing all comments from reviewers yet. will ping you guys once i feel more comfortable with the PR.

whchung commented on third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_rocm.tpl:
```diff
@@ -0,0 +1,239 @@
+  as is as a string to --compiler-options of hipcc. When "-x rocm" is not
+  present, this wrapper invokes gcc with the input arguments as is.
+
+NOTES:
```
ok will do

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+
+def _find_rocm_lib(lib, repository_ctx, cpu_value, basedir, version="",
+                   static=False):
+  """Finds the given CUDA or cuDNN library on the system.
```
thanks. will do.

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+    lib: The name of the library, such as "rocmrt"
+    repository_ctx: The repository context.
+    cpu_value: The name of the host operating system.
+    basedir: The install directory of CUDA or cuDNN.
```
thanks. will do.

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+      return struct(file_name=file_name, path=str(path.realpath))
+
+  elif cpu_value == "Windows":
+    path = repository_ctx.path("%s/lib/x64/%s" % (basedir, file_name))
```
no it's not yet available on Windows. some infrastructure work has been undergoing but it'll be some time before we can actually enable it. I'll remove these checks.

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+  auto_configure_fail("Cannot find rocm library %s" % file_name)
+
+def _find_libs(repository_ctx, rocm_config):
+  """Returns the CUDA and cuDNN libraries on the system.
```
thanks. will fix it.

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+
+  Args:
+    repository_ctx: The repository context.
+    rocm_config: The CUDA config as returned by _get_rocm_config
```
thanks. will fix it.

whchung commented on third_party/gpus/rocm_configure.bzl:
```diff
@@ -0,0 +1,664 @@
+  _tpl(repository_ctx, "rocm:BUILD",
+       {
+           "%{rocmrt_static_lib}": rocm_libs["hip"].file_name,
+           "%{rocmrt_static_linkopt}": '',
```
I'll remove it.

whchung commented on tensorflow/tensorflow.bzl:
```diff
@@ -907,31 +918,53 @@ def _cuda_copts():
                           hdrs=[],
                           **kwargs):
-  copts = copts + _cuda_copts() + if_cuda(cuda_copts) + tf_copts()
+  copts=copts + tf_copts() + _cuda_copts(opts=cuda_copts) + _rocm_copts(opts=cuda_copts)
```
thanks. will fix.

whchung commented on tensorflow/core/kernels/BUILD:
```diff
@@ -2917,7 +2919,9 @@ tf_kernel_library(
+    deps = MATH_DEPS + if_cuda_is_configured([
+        ":cuda_solvers",
+    ]) + if_rocm_is_configured([
         ":cuda_solvers",
```
Unfortunately the header file in `cuda_solvers` target is not CUDA-specific. There are some utility structures used by operators such as `SegmentSumGPUOp` which is supported on ROCm.

I'll submit additional PRs to refactor `SegmentSumGPUOp`. And I'll remove `cuda_solvers` dependency here.

whchung commented on tensorflow/tensorflow.bzl:
```diff
@@ -730,26 +737,27 @@ def tf_cuda_cc_test(name,
-      extra_copts=extra_copts,
-      linkopts=linkopts,
-      args=args)
+  if if_cuda_is_configured(True) or if_rocm_is_configured(True):
```
will do.

whchung commented:
a gentle ping?

whchung commented:
@gunan / @jlebar wondering if you could provide additional inputs to this pull request? There are a few other pull requests I'm preparing depends on the infrastructure established in the PR (ex: `if_rocm_is_configured`, `TENSORFLOW_USE_ROCM`). And some others PRs already filed ( such as #20709 / #20712 / #20715 ) also depends on this PR so they can be further revised / reviewed / merged.

jlebar commented on tensorflow/core/kernels/BUILD:
```diff
@@ -52,7 +52,8 @@ load(
 )
-load("@local_config_cuda//cuda:build_defs.bzl", "if_cuda")
+load("@local_config_cuda//cuda:build_defs.bzl", "if_cuda", "if_cuda_is_configured")
+load("@local_config_rocm//rocm:build_defs.bzl", "if_rocm", "if_rocm_is_configured")
```
> Are we planning to create the internal equivalents of these macros, just like what we have for CUDA?

It's @Artem-B and my opinion that `--config=cuda` (and by corollary `if_cuda`) is a harmful hack.  TF uses it as a switch to say "build for GPU or not", but it shouldn't: This should be decided by BUILD rules.  (E.g. you depend on `:tensorflow_cpu` or `:tensorflow_gpu`.)

The reason we introduced `--config=cuda` is because at the time, Skylark was missing some features we needed in order to make the toolchain work properly.  We think these features are there now.

Unfortunately the direction of these patches -- including the Eigen patch -- sort of doubles down on the notion of --config=cuda.  If we can't build a TF which includes both cuda and rocm bits, then we'll effectively never be able to get rid of --config=cuda.

That's why in XLA we've insisted that we maintain the ability to build for both cuda and rocm, determined by BUILD dependencies.

That said, it's unclear to me what is the alternative to these macros for TF, given that the structure of the rocm Eigen patches does not (?) allow us to build for AMDGPU and NVGPU in the same binary.  So TF may need its own versions of them internally; I don't see how else to do it.

whchung replied:
@jlebar Unfortunately given how GPU common runtime is designed I think it's hard to let TF be configured and built with both CUDA and ROCm at the same time. In `gpu_device.cc`, `EigenCudaStreamDevice` has a compile-time dependency to CUDA constructs. In my current implementation for ROCm, I renamed `EigenCudaStreamDevice` to `EigenGpuStreamDevice` and use `TENSORFLOW_USE_ROCM` macro to switch to ROCm-functional equivalents.

For XLA compiler, it's relatively easy to specify a new set of compiler backend and let it target AMDGPU. But for GPU common runtime, a bigger overhaul might be required if the ultimate goal is to get rid of `--config=cuda`.

That said, I believe such effort to modularize TF runtime, should be deferred to future PRs, after we have better consensus on how that be achieved?

Artem-B replied:
Let me try to clarify the way I'd like to see things working.
The goal is to be able to build any (and all) variants without specifying any extra bazel flags.
I.e. one should be able to say `bazel  //my_project:app_cuda //my_project:app_rocm //my_project:app_cpu` and get all three executables.

Let's suppose the app consists of main.cpp, kernel.cu (for CUDA and ROCm) and kernel_cpu.cpp (for CPU-only). All three app variants would use the same main.o, app_cuda will use kernel-cuda.o built from kernel.cu using CUDA-specific options/defines, kernel-rocm.o from kernel.cu using ROCm-specific options/defines and kernel-cpu would be compiled form kernel_cpu.cpp. User should be able to build any combination of them simultaneously. When you change anything in the build system, there's only one build configuration to test. If you've built an app there's no confusion about what it supports (or does not). I can't count number of times when someone attemted to run CPU-only TF and complained that it does not sees the GPUS or ran CUDA-enabled GPU on machine without GPUs and complained that it failed. Single build configuration  also saves on overall build time as the objects that don't care about CUDA/ROCm will be built only once, instead of once per build config. Considering that CUDA files constitute relatively small subset of tensorflow,  the difference is substantial. 

--config=cuda was inherited from internal Google build and exists for number of reasons that are not relevant to open-source tensorflow. I understand that it is convenient to continue adding extra dimensions to config parameters, but now that we're adding support for another accelerator is the good time to make sure we do it right. Maintaining and debugging multiple build configurations is a royal pain. Having single set of build rules for everything makes things somewhat easier to deal with.

whchung replied:
@Artem-B / @jlebar let me see if i understand this correctly to assess the efforts. It seems you expect one `bazel build` to get executables for all specified targets, and ditch `--config=XXX`.

As a corollary, would the following command be good for you? 1 bazel build get 3 PIP packages:

```
# 1 bazel build get 3 PIP packages
bazel build //tensorflow/tools/pip_package:build_cpu_pip_package //tensorflow/tools/pip_package:build_cuda_pip_package //tensorflow/tools/pip_package:build_rocm_pip_package
```

Also I'm wondering how shall we deal with test targets in `bazel test`?

Artem-B replied:
That's the idea. User tells **what** needs to be built, bazel and whoever maintains the build files figure out **how**.
As for the tests, it should follow roughly the same path.
We currently have rules that automatically generate CPU and GPU test variants already (i.e we generate test_foo for CPU and test_foo_gpu for CUDA), so, in a way, tests should mostly do the right thing already. Running correct set of tests will be similar to what happens now, modulo new suffixes on test targets.

whchung replied:
@gregstren sorry for inviting you to an new discussion thread without prior notification.

I, @whchung, am discussing with @jlebar / @Artem-B / @gunan about how to improve the build system of TensorFlow to make it to support multiple CPU / GPU platforms. And I've discovered your work of `dynamic bazel configurations` [1], `skylark build configurations` [2], and `2018 Bazel Configurability / Multiplatform Roadmap` [3] which eventually led me to bazel roadmaps pages [4] [5].

- [1] https://docs.google.com/document/d/1uoU8t7loTOu6uyzez-ilhcYgEXg4xjViJu0aqrf69TY/edit#
- [2] https://docs.google.com/document/d/1vc8v-kXjvgZOdQdnxPTaV0rrLxtP2XwnD2tAZlYJOqw/edit#
- [3] https://docs.google.com/document/d/1p7pdEwqnLOC-ATl-FdyHBJdk4Ae-5GzAYDvpJeMh4tU/edit?ts=5ac517b9#
- [4] https://www.bazel.build/roadmaps/configuration.html
- [5] https://www.bazel.build/roadmaps/platforms.html

I'm very new to the implementation of `bazel`, let alone how to test / adapt all these upcoming features of it in TensorFlow. I'm wondering could you help share some working example projects with targets tied to different toolchains, so I can learn better from those examples? Thank you very much.

Artem-B replied:
Sorry, I've completely forgot that OSS TF still uses custom crosstool to compile with nvcc. :-(
That's going to be a problem. Crosstool is not something that can be switched per build target,
and NVCC does not support clang as the host compiler. That forces us to have multiple build configurations. We still would be able to do something like `bazel --config=nvcc foo_cuda foo_cpu`, or `bazel foo_rocm foo_cpu` (probably with something like `--config=rocm`) but we will not be able to do all three at once. This complicates things.

I'll need to think about it a bit.

@gregestren -- It sounds like bazel has grown a lot of features lately that I've been missing for CUDA compilation. I'm glad to see that my handwavy proposal for TF compilation seems to be roughly in line with the general direction bazel configurability roadmap seems to head towards. I'll need to take a closer look at the recent changes to get better idea what we can do these days, but it appears that I may have more tools at my disposal than I used to in the past.

gunan replied:
Ok to merge then?

Artem-B replied:
Considering that at the moment we do not have a working alternative, that's probably the least bad option.
It does entangle the build with multiple crosstools, but it's a marginal bump in amount of work we'll need to do in addition to what's needed to deal with --config=nvcc. Making sure rocm build remains working will be a bit of a pain, but I expect bulk of the issues to be figured out on nvcc build, so overall it should not be a major issue.

I'm OK with the patch, but it's ultimately TF team's call.

gregestren replied:
Hi Jack, Artem,

Apologies for my delayed response - I was doing some personal traveling last week.

I'd love to discuss more both your goals and how Bazel's multiplatform changes could help them. We're at a weird state now where lots of new possibilities are opening up (like being able to really support per-target crosstools) but the public APIs are still all coming together. So it's not as simple as "just follow this pattern in the Bazel documentation" but that doesn't mean there aren't options for you.

Since I'm just coming in late into this conversation, I suggest we all get on the same basic page of what's desired. Then we can clarify what features can address your goals, and how well.

Would that work?

bensander commented:
Hi, I work with Jack @whchung here at AMD.    Jack suggested the command-line below - is this what is desired?    So would produce 3 PIPs at the end of the build, but the overall build would be faster since we only compile the common code once?

# bazel build //tensorflow/tools/pip_package:build_cpu_pip_package //tensorflow/tools/pip_package:build_cuda_pip_package //tensorflow/tools/pip_package:build_rocm_pip_package

We are not bazel experts so implementing this looks fairly scary, but perhaps not for someone with proper understanding of the build architecture and access to internal test frameworks.    Can someone provide a reference implementation for the existing targets (CPU, CUDA, etc) - then we could modify this PR to fit into that template?  Or we can treat the unified build steps as a separate feature/PR ?

gunan commented on tensorflow/tensorflow.bzl:
```diff
@@ -17,8 +17,15 @@ load(
     "cuda_default_copts",
 )
+load(
+    "@local_config_rocm//rocm:build_defs.bzl",
```
So do we approve, and start wrestling with this change internally?

gunan commented:
OK, looks like we are ready to start merging this PR. As we also need some internal work on this, it will take some time.
Before I do my final review and approve, could you resolve the final conflicts, one final time?

whchung commented:
@gunan thx. i was out of town yesterday. let me revise the PR one more time so you can take a look at it.

whchung commented:
@gunan resolved the merge conflict. please help review it once again. thanks.

whchung commented:
@gunan I addressed the build errors found by CI. Would you mind re-approve and re-commence the tests? Thanks.

gunan commented:
Looks like all build system seems broken.
Do you have a mac machine you can locally iterate on?
Looks like trying just the CPU build will be useful.

whchung commented:
@gunan I'll try get a Mac to locally reproduce the issue. It seems some of my earlier editions to remove Darwin and Windows from ROCm build path could be related. Since ROCm doesn't support Windows & OSX yet, what would be the best practice to do so?

For Ubuntu it's weird that we are seeing this error:
```
ImportError: libtensorflow_framework.so: cannot open shared object file: No such file or directory
```
In my downstream fork we are always getting this shared lib properly built. I'll need to study it a bit better.

whchung commented:
@gunan I resolved one merge conflict earlier today but all previous build logs were gone. Is there a way to get them?

gunan commented:
All new commits invalidate older builds. I will retrigger them.

drpngx commented:
@yifeif copybara said "An error happened while migrating the change". Trying again.

gunan commented:
@whchung  could you check the build failures?

whchung commented:
@gunan I’m working on this. I’ve already resolved Mac and windows errors on my local boxes. Still haven’t quite sure what’s going on with libtensorflow_framework.so on those Ubuntu targets . I haven’t seen such issues on my end thus far. I’ll revise this PR once I got this figured out

gunan commented:
@whchung looks like it is proving to be too difficult to get everything in this change to work all at once.
I propose breaking this up this way:

1 - only the toolchains in third_party/gpus/rocm*
2 - tensorflow.bzl changes
3 - BUILD file changes
4 - configure and bazel.rc changes
5 - Dockerfile.rocm
6 - build script changes.

it will be 6 PRs, but they will be much easier to work with to get our CI green for both you and us.

whchung commented:
@gunan Just to inform you I'm working on some final touches to this PR and should be able to revise it later today. All failures in previous run should have been addressed. I'll inform you when it's ready.

whchung commented:
@gunan I believe with the new commit I should have addressed all previous build errors encountered on Mac, Windows, and Linux. Could you help restart the test? Thanks.

whchung commented:
@gunan many thanks for executing the test. there is only 1 failure in `MacOS Contrib` target, and from the log i really don't believe it has anything to do with this PR but a timing issue. Would it be possible to re-run that target once again?

whchung commented:
@gunan in `Ubuntu Sanity` there is one error but its log is cryptic to me and again I don't really think it has anything related to the PR itself:

```
=== Sanity check step 3 of 13: do_check_futures_test (Check that python files have certain __future__ imports) ===
Traceback (most recent call last):
  File "check_futures_test.py", line 107, in <module>
    main()
  File "check_futures_test.py", line 103, in main
    raise AssertionError('Error in %s: %s' % (short_path, str(e)))
AssertionError: Error in contrib/lite/tools/accuracy/ilsvrc/generate_validation_labels.py: Missing futures: division absolute_import print_function
```

Neither in this PR nor in TensorFlow `master` branch do we see this `generate_validation_labels.py` so it leads me to believe it's a configuration issue in the CI test node.

gunan commented:
Sanity failure is a known issue, so it did pass all presubmits.
Reviewers, could you take one final look?
@yifeif How should we approach pulling this change into google?

whchung commented:
cc @parallelo for awareness

@gunan / @yifeif , do we expect to see new test targets once this PR gets merged?

gunan commented:
No, we  do not have the hardware, toolchains, license reviews or anything in place for us to be able to build with ROCm. So they definitely wont become blocking presubmits for now.
This, of course will be subject to review later.

In the meantime, we can work with you to setup community supported builds, as outlined here:
https://github.com/tensorflow/community/blob/master/sigs/build/community-builds.md

whchung commented:
@gunan Thanks for sharing the community build page with me. @parallelo would look into it and adapt our CI infrastructure to accommodate that.

Also we do like to revive our other outstanding PRs, specifically those in StreamExecutor and GPU common runtime which are blocked by this particular PR. Also we'll start submitting PRs to enable operators on ROCm which condition upon `TENSORFLOW_USE_ROCM` introduced in this PR. Would it be possible to help expedite allowing this PR be merged? Thanks a lot.

With more developers working on TensorFlow ROCm port now we expect subsequent PRs be revised / maintained in timely fashion.

yifeif commented:
This will need a manual pull for sure @gunan. @whchung do you mind resolving the latest conflicts? Reviewers, let me know if this is ready and I can give pulling a shot.

whchung commented:
@yifeif @gunan, my colleague @deven-amd has debased and modified the PR

dagamayank commented:
@yifeif any update to pull this PR in?

drpngx commented:
@aaroey @meteorcloudy any additional comments?

drpngx commented:
Oh, it looks like we have to import this manually.

gunan commented:
@yifeif is working on the manual import of this, she has been working on this for 2 weeks now.

gunan commented:
We had to revert some "if_cuda_is_configured" uses to if_cuda to make all internal tests to pass, but other than that all of this change has been merged.

yifeif commented:
We finally got this change merged! Thanks for the patience. We needed to change if_cuda_is_configured in tf_cuda_library back to if_cuda to get some internal targets to pass. Feel free to send another PR if this causes any issue and we can work out a patch.

drpngx commented:
Woohoo! Thank you @yifeif !

whchung commented:
Thank you @yifeif and @gunan . We’ll revise other pending PRs for ROCm , as well as submitting new ones :)
